src/docfx/lib/git/GitUtility.cs

```diff
@@ -70,6 +73,129 @@ public static unsafe List<GitCommit>[] GetCommits(string repoPath, List<string>
+        /// Init git repository
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <returns>Task status</returns>
+        public static async Task Init(string cwd)
+        {
+            if (!Directory.Exists(cwd))
+            {
+                Directory.CreateDirectory(cwd);
+            }
+
+            await ExecuteNonQuery(cwd, "init");
+            await ExecuteNonQuery(cwd, @"commit -m ""Init Commit"" --allow-empty");
+        }
+
+        /// <summary>
+        /// Clone git repository from remote to local
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <param name="remote">The remote url</param>
+        /// <param name="path">The path to clone</param>
+        /// <returns>Task status</returns>
+        public static Task Clone(string cwd, string remote, string path)
+        {
+            Debug.Assert(PathUtility.FolderPathHasInvalidChars(path));
+
+            return ExecuteNonQuery(cwd, $"clone {remote} {path.Replace("\\", "/", StringComparison.Ordinal)}", TimeSpan.FromMinutes(20));
+        }
+
+        /// <summary>
+        /// Fetch update from remote
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <returns>Task status</returns>
+        public static Task Fetch(string cwd)
+            => ExecuteNonQuery(cwd, "fetch", TimeSpan.FromMinutes(3));
+
+        /// <summary>
+        /// Pull update from remote
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <param name="remote">The remote name, default is origin</param>
+        /// <returns>Task status</returns>
+        public static Task Pull(string cwd, string remote = null)
+            => ExecuteNonQuery(cwd, $"pull {remote ?? string.Empty}", TimeSpan.FromMinutes(3));
+
+        /// <summary>
+        /// Checkout repo to specified branch
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <param name="create">Create this branch if not exists or not</param>
+        /// <param name="branch">The branch name, default is master</param>
+        /// <returns>Task status</returns>
+        public static Task Checkout(string cwd, bool create, string branch = null)
+            => ExecuteNonQuery(cwd, $"checkout {(create ? "-b" : "")} {branch ?? "master"}", TimeSpan.FromMinutes(5));
+
+        /// <summary>
+        /// Reset current repo to remote branch
+        /// </summary>
+        /// <param name="cwd">The current working directory</param>
+        /// <param name="branch">The remote branch name</param>
+        /// <returns>Task status</returns>
+        public static Task Reset(string cwd, string branch)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(branch));
+
+            return ExecuteNonQuery(cwd, $"reset --hard origin/{branch}", TimeSpan.FromMinutes(10));
```

Comment:
>--hard [](start = 48, length = 6)

this looks dangerous....

Reply:
this is what this reset method do

---
In reply to: [180622259](https://github.com/dotnet/docfx/pull/2611#discussion_r180622259) [](ancestors = 180622259)